041-opa-integration/proposal.md

```diff
@@ -0,0 +1,150 @@
+# Summary
+
+This proposal outlines the beginnings of support for policy enforcement. OPA 
+(Open Policy Agent) will be used as the first policy manager to integrate with.
+
+
+# Motivation
+
+Generally speaking, anything that exposes an HTTP API (whether an individual 
+micro-service or an application as a whole) needs to control who can run those 
+APIs and when. 
+
+Concourse, as a CI platform, also needs to control who can do something and when,
+which can be done by integrating with a policy manager. Before any action, send
+a policy check request, with data that describes the action, to the policy manager,
+and only continue if the policy manager replies a "pass".
+
+Possible policies to apply could be:
+
+* Steps, e.g. code security scan, by certain resource types must be done.
+* Some team's tasks must be run with certain tag.
+* Some docker image registries are not allowed.
+* Some docker images are not not allowed.
+* and so on ...
+
+
+# Proposal
+
+## Support multiple policy managers
+
+Like how multiple credential managers are supported, Concourse should allow other
+policy managers than OPA.
+
+## Policy check points
+
+* Concourse has implemented the audit function. Each action, e.g. `set-pipeline`, 
+generates an audit log. Policy checks can be done right before writing audit logs.
+So that all actions that are audited will also go through policy checks.
```

Comment:
It seems to me that this RFC has a pre-requisite that Concourse first expand the simplistic auditing capabilities that were added in 5.2 - documenting explicitly what is audited, a JSON schema of generated audit logs, etc.

Currently, auditing is configured through boolean environment variables. Maybe this should be changed? What if a policy requires an audit which is not enabled through the environment variables - should the action automatically be denied or permitted, should it be impossible to apply the policy to begin with?

It seems to me that there should first be an RFC on maturing auditing before putting work in on policy engines.

Reply:
Thanks for the comment and sorry for the confusion. The policy check doesn't depend on audit at all. Let me fix the wording.